nack/receive_log.go

```diff
@@ -0,0 +1,130 @@
+package nack
+
+import (
+	"errors"
+	"strconv"
+)
+
+var (
+	allowedReceiveLogSizes     map[uint16]bool
+	invalidReceiveLogSizeError string
+)
+
+func init() {
+	allowedReceiveLogSizes = make(map[uint16]bool, 15)
+	invalidReceiveLogSizeError = "invalid ReceiveLog size, must be one of: "
+	for i := 6; i < 16; i++ {
+		allowedReceiveLogSizes[1<<i] = true
+		invalidReceiveLogSizeError += strconv.Itoa(1<<i) + ", "
+	}
+	invalidReceiveLogSizeError = invalidReceiveLogSizeError[0 : len(invalidReceiveLogSizeError)-2]
+}
+
+type ReceiveLog struct {
+	packets         []uint64
+	size            uint16
+	end             uint16
+	started         bool
+	lastConsecutive uint16
+}
+
+func NewReceiveLog(size uint16) (*ReceiveLog, error) {
+	if !allowedReceiveLogSizes[size] {
+		return nil, errors.New(invalidReceiveLogSizeError)
+	}
+
+	return &ReceiveLog{
+		packets: make([]uint64, size/64),
```

Comment:
This should be an interface: if my application is already logging packets for some other reason, then I should be able to pass my receive log to the interceptor rather than keeping yet another copy of the log.

Reply:
Buffers should not be shared. Here is my reasoning:

receive_log:
Receive log stores only the fact that the packet arrived. That means it uses 1 bit per packet, for 8192 pacekts, it uses 1024bytes. If I remember correctly the mtu is about 1.5 kB, so that means the whole receive log is smaller than a single video packet.

send_buffer:
This is more tricky, because it stores every packet. If you have multiple outgoing stream, like in the case of a SFU, it will store the packets for every stream. But:
- packet payload is a slice, meaning the underlying data is not copied, it's only stored once
- header will be mostly different for every sfu client, because payload type, extensions, ssrc, etc. can differ
If reusing a buffer, you would have to fix all headers every time you send out a packet. This might work for an sfu server, but there are more complicated use cases than that.
- in a centralized buffer it's much harder to configure when a packet can be discarded (eg: in the future the nack sender interceptor can listen to tcc packets as well, allowing it to drop packets based on it)

The final reason I'm against a single shared buffer is jitter handling. You might want to use the same stream with and without a jitter buffer at the same time. We have an aplication that saves a stream to a webm file on disk, but also forwards the stream to a single viewer. We then proxy the tcc packets from the viewer back to the broadcaster. In this case, the file save part needs a jitter buffer, but the proxy part does not, because the jitter buffer messes with the packet delays from the viewer's perspective, and messes up the whole tcc.

Reply:
Sorry, yes, I meant the send buffer.

In this comment, I'm not arguing against or in favour of a shared buffer.  I'm merely arguing that the send buffer should be an interface that the user implements.  Your current implementation is probably fine as the default, but the user should be able to provide a different implementation for the buffer.